### ISSUE TITLE
recv() not flexible enough

### ISSUE THREAD

**Issue by only-a-ptr** [Ybalrid/kissnet#5]
Hello.

I'm porting a [TCP-based game server](https://github.com/RigsOfRods/ror-server) to kissnet and I'm struggling with recv() being limited to constant-sized buffer:
* since OS recv() receives data _up to_ the buffer size, the `kissnet::buffer<>` can always end up half empty and there's no way to append to it.
* if the message size isn't known in advance (as in my case), I must either use a large enough constant-size buffer or dynamically allocate appropriate size.

I tried coding a 'receive up to buffer capacity' function on top of current kissnet's recv(), but I ended up with a monster: https://github.com/only-a-ptr/ror-server/blob/5e081702f540c0c811e3a7658b4496d3677d3750/source/server/messaging.cpp#L290-L328

I see many possible solutions to this:
* Smarter buffer with embedded size counter - much like STL containers but with constant-size capacity; not complete solution but simple, stupid!
* Start index as parameter `recv(kissnet::buffer buf, size_t start)`
* Iterators `recv(kissnet::buffer::iterator start, kissnet::buffer::iterator end)`
* Pointer arithmetic `recv(std::byte* buf, size_t len)`
* new func `recv_all()` which repeats OS recv() until buffer is full - not complete solution but simple, stupid!

**Comment by Ybalrid** [comment]
I'll have to think about that 😄

**Comment by only-a-ptr** [comment]
I noticed send() is overloaded: the central one is plain old `(byte* buffer, size_t len)` and the other is a wrapper using `kissnet::buffer`?

Maybe just remake recv() in the same manner? That way it'd be consistent and you wouldn't even need to update tutorials 😉

**Comment by Ybalrid** [comment]
Yes, whatever the solution is, it will be an overload of recv. I'm going to add something equally simple and stupid to what's been done with send.

My preferred solution in that case would just the "pointer and size" method anyway. It's just a few lines to write, if you happy to do it you open a pull request if I forgot to add it this weekend :wink:

**Comment by Ybalrid** [comment]
By the way, this is super cool that you are using kissnet for Rigs of Rod's netcode! Can you tell me more about the how and why you decided to switch socket libraries? It could give me some insight into what people want to use such a library.

If you are curious (you may have read the blog articles linked in README) I wrote kissnet quite quickly after being quite frustrated by the un-necessary complexity of C++ networking libraries I could find, and especially Boost.Asio. And especially because the "Networking TS", a thing bound to be merged into the standard C++ library one day is based on it, and it contains a lot of stuff that I believe most people don't need in a library that represent a socket.

My take is that a "socket" class should be a easy projection of the Operating System-level concept of what a "socket" is : an opaque object that represent a place where I can get bytes from another computer, and send these bytes to that other computer.

**Comment by Ybalrid** [comment]
@only-a-ptr Master now have an `recv` that only takes a pointer and a size 😉 

Is this good enough to solve your issue? I don't want to add "data buffer management code" to kissnet, nor a "packet" or "variable buffer" abstraction, all of that is the responsibility of the client code.

I may add a "offset" argument (defaulted to zero) to the `recv` that takes a "buffer" too, that can be useful

**Comment by only-a-ptr** [comment]
> @Ybalrid Master now have an recv that only takes a pointer and a size 😉

I was just in the middle of writing a nicer `RecvAll()` wrapper myself 😄 Pointer+size is eactly what I wished for, thanks!

> @Ybalrid By the way, this is super cool that you are using kissnet for Rigs of Rod's netcode!

It's just survival, really 😆 We use https://github.com/RigsOfRods/socketw which we maintain out of necessity, it's been abandoned since ages ago. Kissnet is the only drop-in replacement I ever found, for the longest time I looked at http://enet.bespin.org/. I only learned how sockets work recently and I realized the server could be way easier to maintain with non-blocking sockets. I didn't trust SocketW to handle the switch, in fact I have zero trust towards it, so I ported right away.

**Comment by Ybalrid** [comment]
All of this is really nice to hear!

I'll let you close this issue if everything's fine for you regarding `recv` now.

(You also can do kissnet::buffer + "data start offset" too now.)

Sockets are fun, and it's like threads, people make them look scarier than they should be.

Looking at socketw, the only feature I think I do not have here is SSL/TLS, but again, I don't really think I should provide that directly... Or maybe I could provide some interfaces to simplify the usage of kissnet with OpenSSL or LibreSSL... But I fear that's starting to open a crazy can of worms :joy:

In any case, once you merge the kissnet code into the main project, do not hesitate to add Pull Request here a new section in the README.md "Projects using Kissnet" and start a list. I am aware of one more project that started using this library, but I think it's nice to have a little list like that.

**Comment by only-a-ptr** [comment]
I can close now: https://github.com/RigsOfRods/ror-server/pull/98

BTW yes I did read https://blog.ybalrid.info/2018/12/no-nonsense-networking-for-c-introducing-kissnet-a-k-i-s-s-socket-library/ and I second it wholeheartedly. I've sticked to SocketW all this time for pretty much the same reasons, to the point I was almost ready to start maintaining it. ASIO feels a lot like rorserver in it's current state: the author spinned a pile of (poorly synced) threads to wait for blocking sockets instead of learning to use nonblocking sockets.

I'll be in touch, come visit to RoR forums or Discord if you dare 😉
